packager/media/test/test_web_server.cc

```diff
@@ -0,0 +1,261 @@
+}
+
+bool is_mg_str_null(const mg_str& mg_string) {
+  return mg_string.ptr == NULL;
+}
+
+bool is_mg_str_null_or_blank(const mg_str& mg_string) {
+  return mg_string.ptr == NULL || mg_string.len == 0;
+}
+
+// Get a string query parameter from a mongoose HTTP message.
+bool get_string_query_parameter(struct mg_http_message* message,
+                                const char* name,
+                                std::string_view* str) {
+  struct mg_str value_mg_str = mg_http_var(message->query, mg_str(name));
+
+  if (!is_mg_str_null(value_mg_str)) {
+    *str = view_mg_str(value_mg_str);
+    return true;
+  }
+
+  return false;
+}
+
+// Get an integer query parameter from a mongoose HTTP message.
+bool get_int_query_parameter(struct mg_http_message* message,
+                             const char* name,
+                             int* value) {
+  std::string_view str;
+
+  if (get_string_query_parameter(message, name, &str)) {
+    return absl::SimpleAtoi(str, value);
+  }
+
+  return false;
+}
+
+}  // namespace
+
+namespace shaka {
+namespace media {
+
+TestWebServer::TestWebServer() : status_(kNew), stopped_(false) {}
+
+TestWebServer::~TestWebServer() {
+  {
+    absl::MutexLock lock(&mutex_);
+    stop_.Signal();
+    stopped_ = true;
+  }
+  if (thread_) {
+    thread_->join();
+  }
+  thread_.reset();
+}
+
+bool TestWebServer::Start(int port) {
+  thread_.reset(new std::thread(&TestWebServer::ThreadCallback, this, port));
+
+  absl::MutexLock lock(&mutex_);
+  while (status_ == kNew) {
+    started_.Wait(&mutex_);
+  }
+
+  return status_ == kStarted;
+}
+
+void TestWebServer::ThreadCallback(int port) {
+  // Mongoose needs an HTTP server address in string format.
+  // "127.0.0.1" is "localhost", and is not visible to other machines on the
+  // network.
+  std::string http_address = absl::StrFormat("http://127.0.0.1:%d", port);
+
+  // Set up the manager structure to be automatically cleaned up when it leaves
+  // scope.
+  std::unique_ptr<struct mg_mgr, decltype(&mg_mgr_free)> manager(
+      new struct mg_mgr, mg_mgr_free);
+  // Then initialize it.
+  mg_mgr_init(manager.get());
+
+  auto connection =
+      mg_http_listen(manager.get(), http_address.c_str(),
+                     &TestWebServer::HandleEvent, this /* callback_data */);
+  if (connection == NULL) {
+    // Failed to listen to the requested port.  Mongoose has already printed an
+    // error message.
+    absl::MutexLock lock(&mutex_);
+    status_ = kFailed;
+    started_.Signal();
+    return;
+  }
+
+  {
+    absl::MutexLock lock(&mutex_);
+    status_ = kStarted;
+    started_.Signal();
+  }
+
+  bool stopped = false;
+  while (!stopped) {
+    // Let Mongoose poll the sockets for 100ms.
+    mg_mgr_poll(manager.get(), 100);
+
+    // Check for a stop signal from the test.
+    {
+      absl::MutexLock lock(&mutex_);
+      stopped = stopped_;
+    }
+  }
+
+  {
+    absl::MutexLock lock(&mutex_);
+    status_ = kStopped;
+  }
```

Comment:
nit: why not just do it after line 134,

```
if (stopped)
  status_ = kStopped;
```

Avoids one extra lock.

Reply:
Ah, good point.  Done.